server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java

```diff
@@ -78,31 +82,51 @@ protected ReplicationResponse newResponseInstance(StreamInput in) throws IOExcep
     protected void shardOperationOnPrimary(
         BasicReplicationRequest shardRequest,
         IndexShard primary,
-        ActionListener<PrimaryResult<ReplicaShardRefreshRequest, ReplicationResponse>> listener
+        ActionListener<PrimaryResult<BasicReplicationRequest, ReplicationResponse>> listener
     ) {
-        ActionListener.completeWith(listener, () -> {
+        try (
+            var listeners = new RefCountingListener(
+                listener.map(ignored -> { return new PrimaryResult<>(shardRequest, new ReplicationResponse()); })
+            )
+        ) {
             var refreshResult = primary.refresh(SOURCE_API);
             logger.trace("{} refresh request executed on primary", primary.shardId());
-            var shardRefreshRequest = new ReplicaShardRefreshRequest(
-                primary.shardId(),
-                shardRequest.getParentTask(),
-                refreshResult.generation()
-            );
-            return new PrimaryResult<>(shardRefreshRequest, new ReplicationResponse());
-        });
+
+            // Forward the request to all nodes that hold unpromotable replica shards
+            final ClusterState clusterState = clusterService.state();
+            clusterService.state()
+                .routingTable()
+                .shardRoutingTable(shardRequest.shardId())
+                .replicaShards()
+                .stream()
+                .filter(Predicate.not(ShardRouting::isPromotableToPrimary))
+                .map(ShardRouting::currentNodeId)
+                .collect(Collectors.toUnmodifiableSet())
+                .forEach(nodeId -> {
+                    final DiscoveryNode node = clusterState.nodes().get(nodeId);
+                    UnpromotableReplicaRefreshRequest request = new UnpromotableReplicaRefreshRequest(
+                        primary.shardId(),
+                        refreshResult.generation()
+                    );
+                    logger.trace("forwarding refresh request [{}] to node [{}]", request, node);
+                    transportService.sendChildRequest(
+                        node,
+                        TransportUnpromotableReplicaRefreshAction.NAME,
+                        request,
+                        taskManager.getTask(shardRequest.getParentTask().getId()),
+                        TransportRequestOptions.EMPTY,
+                        new ActionListenerResponseHandler<>(listeners.acquire(ignored -> {}), (in) -> TransportResponse.Empty.INSTANCE)
```

Comment:
We will handle this response on a `transport_worker` thread, which means we'll complete the outer listener on such a thread too. Is that ok? Maybe we should fork back to a `REFRESH` thread at the end just to be safe.

Reply:
Hi @DaveCTurner ! Nice catch. I'll add the thread pool argument to the response handler here.